+act #15819 Create ask() variant that can be used without sender(); Add PromiseRef as an alternative to ask pattern

ataraxer opened the pull request:
Closes #15819

ktoso commented:
OK TO TEST

ataraxer commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -207,53 +255,106 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
  */
-private[akka] final class PromiseActorRef private (val provider: ActorRefProvider, val result: Promise[Any], _mcn: String)
-  extends MinimalActorRef {
+private[akka] final class PromiseActorRef(val provider: ActorRefProvider)
```
This should probably be renamed to `PromiseRefImpl` for consistency with Akka Typed code and to be less ambiguous with public `PromiseRef`. However I've found mentions of `PromiseActorRef` in documentation and it is also used in public `ActorPromiseActorRef` java API (although I couldn't figure out what's it purpose or how is it supposed to be used) so I've decided to leave it unchanged for now.

ataraxer commented:
Tests failed with bianry compatibility issues. Reproduced locally:
```
akka > akka-actor/mimaReportBinaryIssues
[info] Resolving com.typesafe#config;1.2.1 ...
[info] akka-actor: found 2 potential binary incompatibilities (filtered 238)
[error]  * method apply(akka.actor.ActorRefProvider,akka.util.Timeout,java.lang.String)akka.pattern.PromiseActorRef in object akka.pattern.PromiseActorRef does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.PromiseActorRef.apply")
[error]  * method this(akka.actor.ActorRefProvider,scala.concurrent.Promise)Unit in class akka.pattern.PromiseActorRef does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.PromiseActorRef.this")
[trace] Stack trace suppressed: run last akka-actor/*:mimaReportBinaryIssues for the full output.
[error] (akka-actor/*:mimaReportBinaryIssues) akka-actor: Binary compatibility check failed!
[error] Total time: 0 s, completed Aug 18, 2015 10:18:29 AM
```

`PromiseActorRef` is internal API -- added reported issues to exceptions:
```scala
ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.PromiseActorRef.apply"),
ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.PromiseActorRef.this")
```

```
akka > akka-actor/mimaReportBinaryIssues
[info] Resolving com.typesafe#config;1.2.1 ...
[info] akka-actor: found 0 potential binary incompatibilities (filtered 240)
[success] Total time: 1 s, completed Aug 18, 2015 10:24:08 AM
```

ktoso commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -396,29 +497,39 @@ private[akka] final class PromiseActorRef private (val provider: ActorRefProvide
+    if (timeout.duration.length > 0) {
+      new PromiseActorRef(provider).setTimeout(timeout)
+    } else {
+      throw new IllegalArgumentException(s"Timeout length must not be negative.")
```
Include the timeout value, like: `", was $timeout."`

ktoso commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -73,8 +76,21 @@ trait AskSupport {
+   *   ask(worker, respondTo => Message(respondTo))(timeout)
+   * }}}
+   */
+  def ask(actorRef: ActorRef, message: ActorRef ⇒ Any)(implicit timeout: Timeout): Future[Any] =
```
This introduces a tricky question about Java API:
- We should provide Java API for it, since on Scala 2.11 it's not nice to use this ask, it would have to be `ask(ActorRef, japi.Function1)`
- Scala 2.12 comes out, and `ActorRef => Any` can now be used from Java 8 as `e -> m`, won't java8 freak out that it's ambiguous which method we wanted - the "java api" (added for Scala 2.11 + JDK7) or the "scala api, which now works with Java 8"?

So... Since 2.4 requires JDK8, Java users will most likely be able to bump from 2.11 to 2.12 since they don't depend on Scala compat (I'd hope), and use the new API via the scala Functions. Summing up - it's best we do not add specific Java API for this method? Unless I missed something.

ataraxer replied:
I think it would be better to provide `ask(ActorRef, japi.Function)` now and address this problem after 2.12 release, so that `japi.Function` replacement with lambdas can be solved in a more uniform way across all of Java APIs. A single case of API which provides lambdas interop would introduce inconsistency, which is worse than potential ambiguity.

ataraxer commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -150,36 +180,54 @@ final class AskableActorRef(val actorRef: ActorRef) extends AnyVal {
+    internalAsk(message, timeout, ActorRef.noSender)
+
+  def ?(message: ActorRef ⇒ Any)(implicit timeout: Timeout, sender: ActorRef): Future[Any] =
     internalAsk(message, timeout, sender)
```
Unfortunately it seems that this won't work. We can't provide a single signature `?(ActorRef => Any)(implicit Timeout, ActorRef)`, because multiple overloads with default arguments are not allowed.

On the other hand current implementation fails miserably with "ambiguous reference to overloaded definition", due to erasure, which means that we need to choose to leave either `?(ActorRef => Any)(implicit Timeout, ActorRef)` or `?(ActorRef => Any)(implicit Timeout)`.

First option would require to explicitly specify an implicit sender, adding some cruft to using ask pattern from non actor context (which is one of the primary use cases of the pattern):
```scala
implicit val sender = ActorRef.noSender
target ? (Request(_))
```

Second one will make is impossible to pass a reference to sender.

ataraxer replied:
Other options:
- provide a separate import for this use-case as a `richAsk` or `extendedAsk`;
- choose a different name:
  - `??` -- too close to `???`;
  - `>?<` -- takes longer to write;
  - `=?>` -- welcome to "this is why nobody likes symbolic operators" zone.

This also raises a question of consistency of usage across `akka-actor` and `akka-typed`, considering that the latter already provides a `?(Any => ActorRef)(Timeout)` option.

cc @rkuhn: any thoughts?

rkuhn replied:
We should not introduce a new symbol—remember `!!` and `!!!`?—and overloading will not work (even in those cases where it does work, type inference will fail for the function literal), so providing it as another import is the only option. Thanks for the exploration!

BTW: the symmetry with `akka-typed` is the reason for doing this, explicit replyTo address style is not only useful in that context but also here—implicit `sender` is sometimes just a tad too hidden.

ktoso commented:
FYI, this won't make it into the RC1 release (we're cutting it tomorrow). Let's keep the discussion going though, it's a quite involved feature which we should "get *right*" before pulling in.

hepin1989 commented:
@ktoso This feature would be great if it would be inclouded in akka2.4.x.but I can see that akka team is rushing:)

ktoso commented:
It can be in `2.4.x`, it just won't make it for the `RC1`.

ataraxer commented:
OK, I think I'm done for now:
- moved new API to `akka.pattern.extended` package which can aggregate other conflicting pattern APIs in future; this also ensures that the API is backward compatible in case someone uses a wildcard import like `import akka.pattern._`;
- added Java API for both Scala 2.11 and 2.12 options so it can be used immediately after 2.12 release;
- added tests for `ask` and `expanded.ask`;
- added tests for `PromiseRef`;
- cleaned up the docs;

ataraxer commented:
Ping?

rkuhn commented on akka-actor-tests/src/test/scala/akka/pattern/AskSpec.scala:
```diff
@@ -1,7 +1,8 @@
  */
-package akka.pattern
+/* Wrapper package has to be added in order to prevent implicits from clashing. */
+package akka.pattern.wrapper
```
This file does not add implicits, so this comment seems incorrect; otherwise this file would now also be in the wrong source directory.

rkuhn commented on akka-actor/src/main/scala/akka/pattern/package.scala:
```diff
@@ -40,6 +40,51 @@ import akka.actor._
+   * });
+   * }}}
+   */
+  package object extended extends ExplicitAskSupport
```
This should not be a package object and it should also not be within the package object for `akka.pattern`—package objects are strange beasts. I think this should be a normal `object` within the `akka.pattern` package.

rkuhn commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -73,9 +192,9 @@ trait AskSupport {
    *
    */
   def ask(actorRef: ActorRef, message: Any)(implicit timeout: Timeout): Future[Any] =
-    actorRef.internalAsk(message, timeout, ActorRef.noSender)
```
these changes should probably not be done: the old ask support was the most efficient solution for that problem, calling internalActorAsk means creating one closure per call that is then not really used (i.e. it only transports the final value anyway)

rkuhn commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -150,38 +270,31 @@ final class AskableActorRef(val actorRef: ActorRef) extends AnyVal {
-        a.result.future
-      }
-    case _ ⇒ Future.failed[Any](new IllegalArgumentException(s"""Unsupported recipient ActorRef type, question not sent to [$actorRef]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
-  }
```
This method must remain: it is needed for binary compatibility.

rkuhn commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -189,13 +302,13 @@ object AskableActorSelection {
    * INTERNAL API: for binary compatibility
    */
   private[pattern] def ask$extension(actorSel: ActorSelection, message: Any, timeout: Timeout): Future[Any] =
-    actorSel.internalAsk(message, timeout, ActorRef.noSender)
```
these changes should also be removed

rkuhn commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -207,53 +320,153 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
+
+  /**
+   * INTERNAL API: for binary compatibility
+   */
```
for new classes there can be no BC concerns: only add those methods that are really needed and comment them meaningfully

rkuhn commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -207,53 +320,153 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
+/**
+ * INTERNAL API: for binary compatibility
+ */
+private[pattern] object InternalAsk {
```
the methods in here should be moved into the ExplicitlyAskableActorRef/Selection, respectively

rkuhn commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -207,53 +320,153 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
     case ref: InternalActorRef ⇒
-      if (timeout.duration.length <= 0)
+      if (timeout.duration.length <= 0) {
+        val message = messageFactory(ActorRef.noSender)
```
This is just a glorified `null`; we should probably use `system.deadletters` instead (to give a real ActorRef).

rkuhn replied:
(same below)

rkuhn commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -207,53 +320,153 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
  *
  * INTERNAL API
  */
-private[akka] final class PromiseActorRef private (val provider: ActorRefProvider, val result: Promise[Any], _mcn: String)
```
I think this class will need to remain binary compatible (it is used by prominent “naughty” projects).

rkuhn commented on akka-actor/src/main/scala/akka/pattern/Patterns.scala:
```diff
@@ -48,6 +48,37 @@ object Patterns {
+   * A variation of ask which allows to implement "replyTo" pattern by including
+   * sender reference in message.
+   *
+   * For use with Akka compiled with Scala 2.12.
```
Hmm, it would be nice to just do “the right thing” for each release, but that might be impossible. Given that `japi.Function` works fine independent of Scala version I’d remove this variant without replacement—in a different universe this would be the only one we need …

rkuhn commented:
Sorry for the delay, this fell through the cracks :-( Overall it looks promising but I do not yet understand why the internals were changed, both for untyped and typed. Especially in the untyped case our hands are pretty tied because this was such a useful feature that people were naughty and started using internal APIs; if we can avoid it then we should definitely not break binary compatibility.

ataraxer committed:
+act #15819 Create ask() variant that can be used without sender()

ataraxer committed:
+act #15819 Add PromiseRef and FutureRef as an alternative to ask pattern

ataraxer commented:
Sorry for the delay on my part too -- had a bit of vacation time myself :)

I had reworked the original implementation from the ground up effectively turning it into an addition only PR. Binary compatibility FTW :)

"Explicit sender" pattern is now explicitly enabled via `akka.pattern.extended.ask` import and is implemented separately from the original ask to avoid changing the code in breaking ways.

`PromiseRef` has been reimplemented as a union of `ActorRef` and `Promise` represented as class members instead of type-level union (which turned out to be quite tricky to implement and fragile to changes) with optional `Implicits` provided to make it's use transparent wherever `ActorRef` or `Promise` is applicable.

`FutureRef` has also been added to mirror the `Promise/Future` relationship and add some flexibility to the further application of `FutureRef` and `PromiseRef`, which turned out to be really useful abstractions on themselves, enabling to express one-time computations performed by an actor which are exposed in a type-safe level as a value of a `Promise` instead of a message dispatch.

ataraxer commented on akka-actor/src/main/scala/akka/pattern/extended/package.scala:
```diff
@@ -0,0 +1,49 @@
+ * import static akka.pattern.Patterns.ask;
+ *
+ * // for Akka compiled with Scala 2.12
+ * ask(actor, askSender -> new Request(askSender));
```
This comment will be removed: leftover from the original PR.

ataraxer commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -184,6 +300,42 @@ final class AskableActorRef(val actorRef: ActorRef) extends AnyVal {
+        a.result.future
+      }
+    case _ ⇒
+      val message = messageFactory(sender.asInstanceOf[InternalActorRef].provider.deadLetters)
```
I'm a bit concerned about this use of `deadLetters`, which is inconsistent from those defined for `InternalActorRef`s above. In the case of local actor communication within the same `ActorSystem` it will be essentially the same mailbox as in `InternalActorRef`s case. However with cross-system communication and/or remoting it becomes a `deadLetters` of sender's system, instead of remote one, which seems at the very least strange semantically. Is it something to be concerned about? Maybe it should be replaced with some kind of dummy `ActorRef` which does nothing at all.

On the other hand, if provided reference is not an `InternalActorRef` there can be no dispatch at all, so in some sense this letter can only be "dead" for the local system... Does this makes sense?

rkuhn replied:
Using deadLetters like this is the right thing to do, it is consistent with the rest of Akka. However, this line will throw an NPE if no sender was given, so please check that case and bail out without invoking the function in this case.

drewhk commented:
I personally like this. @rkuhn ?

rkuhn commented on akka-actor/src/main/scala/akka/pattern/PromiseRef.scala:
```diff
@@ -0,0 +1,195 @@
+     * Provides a conversion to transparently use PromiseRef to dispatch messages via Scala DSL.
+     */
+    implicit def toScalaActorRef(promiseRef: PromiseRef[_]): ScalaActorRef = promiseRef.ref
+  }
```
I’m not sure about these, we normally don’t do that. At the cost of one more import these conversions make the code one bit less obvious in order to save a few keystrokes—unless of course I’m missing something.

rkuhn commented on akka-actor/src/main/scala/akka/pattern/PromiseRef.scala:
```diff
@@ -0,0 +1,195 @@
+     * Provides a conversion to transparently use FutureRef to dispatch messages via Scala DSL.
+     */
+    implicit def toActorRef(futureRef: FutureRef[_]): ActorRef = futureRef.ref
+  }
```
same here; and if these should go then the language import should go with them

rkuhn commented on akka-actor/src/main/scala/akka/pattern/PromiseRef.scala:
```diff
@@ -0,0 +1,195 @@
+private[akka] object AskPromiseRef {
+  def apply(provider: ActorRefProvider, timeout: Timeout): AskPromiseRef = {
+    if (timeout.duration.length > 0) {
+      val promiseActorRef = PromiseActorRef(provider, timeout, "unknown", "unknown", ActorRef.noSender)
```
please use `provider.deadLetters` instead of `null` (which is what `noSender` means)

rkuhn commented:
Thanks for the updates @ataraxer , this looks pretty good!

ataraxer commented:
Updated and rebased to master.

rkuhn commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -238,6 +392,41 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
+    case null ⇒
+      Future.failed[Any](new IllegalArgumentException(s"""No recipient provided, question not sent to [$actorSel]."""))
+    case _ ⇒
+      val message = messageFactory(sender.asInstanceOf[InternalActorRef].provider.deadLetters)
```
sorry for being unclear: it is the `sender` that can be `null`, not the `ref`

ataraxer replied:
Oh, of course, my bad. Rushed it a bit.

rkuhn commented:
Thanks @ataraxer !

ataraxer commented:
Thanks for guiding me through my first contribution to Akka! It's been a lovely experience (:

rkuhn commented:
Great, you can always come back for more :-)

ktoso commented:
Wohoo, awesome this made it in! Thanks @ataraxer :-)
